src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,278 @@
+        //   of [IDNA] notwithstanding). If the server needs to match the HostName
+        //   against names that contain non-US-ASCII characters, it MUST perform
+        //   the conversion operation described in section 4 of [IDNA], treating
+        //   the HostName as a "query string" (i.e. the AllowUnassigned flag MUST
+        //   be set). Note that IDNA allows labels to be separated by any of the
+        //   Unicode characters U+002E, U+3002, U+FF0E, and U+FF61, therefore
+        //   servers MUST accept any of these characters as a label separator.  If
+        //   the server only needs to match the HostName against names containing
+        //   exclusively ASCII characters, it MUST compare ASCII names case-
+        //   insensitively.
+        private static string GetSniFromServerNameList(ReadOnlySpan<byte> serverNameListExtension, out ReadOnlySpan<byte> remainingBytes)
+        {
+            if (serverNameListExtension.Length < 2)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            int serverNameListLength = ReadUint16(serverNameListExtension);
+            ReadOnlySpan<byte> serverNameList = serverNameListExtension.Slice(2);
+
+            if (serverNameListLength > serverNameList.Length)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            remainingBytes = serverNameList.Slice(serverNameListLength);
+            ReadOnlySpan<byte> serverName = serverNameList.Slice(0, serverNameListLength);
+
+            if (serverName.Length < 3)
+            {
+                return null;
+            }
+
+            // -1 for hostNameType
+            int hostNameStructLength = ReadUint16(serverName) - 1;
+            byte hostNameType = serverName[2];
+            ReadOnlySpan<byte> hostNameStruct = serverName.Slice(3);
+
+            if (hostNameStructLength != hostNameStruct.Length || hostNameType != 0x00)
+            {
+                return null;
+            }
+
+            int hostNameLength = ReadUint16(hostNameStruct);
+            ReadOnlySpan<byte> hostName = hostNameStruct.Slice(2);
+
+            return DecodeString(hostName);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static string DecodeString(ReadOnlySpan<byte> bytes)
+        {
+            string idnEncodedString = Encoding.UTF8.GetString(bytes);
+            try
+            {
+                return s_idnMapping.GetUnicode(idnEncodedString);
+            }
+            catch (ArgumentException)
+            {
+                // client has not done IDN mapping
+                return idnEncodedString;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int ReadUint16(ReadOnlySpan<byte> bytes)
+        {
+            return (bytes[0] << 8) | bytes[1];
```

Comment:
BinaryPrimitives.ReadUInt16BigEndian has a different implementation than this, which we theoretically think is faster.  You should probably just call it.